dask/dataframe/io/parquet/core.py

```diff
@@ -711,81 +781,68 @@ def to_parquet(
+            raise ValueError("``name_function`` must produce unique filenames.")
 
-    if name_function is not None and len(set(filenames)) < len(filenames):
-        raise ValueError("``name_function`` must produce unique filenames.")
-
-    # Construct IO graph
-    dsk = {}
-    name = "to-parquet-" + tokenize(
-        df,
-        fs,
-        path,
-        append,
-        ignore_divisions,
-        partition_on,
-        division_info,
-        index_cols,
-        schema,
-    )
-    part_tasks = []
+    # Create Blockwise layer for parquet-data write
     kwargs_pass["fmd"] = meta
     kwargs_pass["compression"] = compression
     kwargs_pass["index_cols"] = index_cols
     kwargs_pass["schema"] = schema
-    for d, filename in enumerate(filenames):
-        dsk[(name, d)] = (
-            apply,
-            engine.write_partition,
-            [
-                (df._name, d),
-                path,
-                fs,
-                filename,
-                partition_on,
-                write_metadata_file,
-            ],
-            toolz.merge(kwargs_pass, {"head": True}) if d == 0 else kwargs_pass,
-        )
-        part_tasks.append((name, d))
-
-    final_name = "metadata-" + name
-    # Collect metadata and write _metadata
-
-    if write_metadata_file:
-        dsk[(final_name, 0)] = (
-            apply,
-            engine.write_metadata,
-            [
-                part_tasks,
-                meta,
-                fs,
-                path,
-            ],
-            {"append": append, "compression": compression},
-        )
-    else:
-        dsk[(final_name, 0)] = (lambda x: None, part_tasks)
-
-    graph = HighLevelGraph.from_collections(final_name, dsk, dependencies=[df])
+    data_write = df.map_partitions(
+        ToParquetFunctionWrapper(
+            engine,
+            path,
+            fs,
+            partition_on,
+            write_metadata_file,
+            i_offset,
+            name_function,
+            kwargs_pass,
+        ),
+        BlockIndex((df.npartitions,)),
+        meta=df._meta,
+        enforce_metadata=False,
+        transform_divisions=False,
+        align_dataframes=False,
+    )
 
+    # Convert data_write + metadata write to computable collection
+    meta_name = "metadata-" + data_write._name
+    graph = HighLevelGraph.from_collections(
+        meta_name,
+        DataFrameTreeReduction(
+            name=meta_name,
+            name_input=data_write._name,
+            npartitions_input=data_write.npartitions,
+            concat_func=lambda x: x,
+            split_every=data_write.npartitions,
+            tree_node_func=lambda x: x,
+            finalize_func=partial(
+                engine.write_metadata,
+                meta=meta,
+                fs=fs,
+                path=path,
+                append=append,
+                compression=compression,
+            )
+            if write_metadata_file
+            else lambda x: None,
```

Comment:
We don't really "need" this to be a tree reduction (especially without defining the concat_func), but I felt it was worth making the minimal changes to move the entire `to_parquet` call to HLG.

Reply:
I've decided to roll back this change and save it for a distinct PR (focused only on parquet-metadata writing).

Reply:
Yeah, it's much harder to read as a tree reduction layer. Maybe a `tree_reduce` helper function for DataFrames would be worthwhile (`aca` could call into this too)?